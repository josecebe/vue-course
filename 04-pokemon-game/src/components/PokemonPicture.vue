<template>
    <div class="pokemon-container">
        <img alt="Pokémon" class="hidden-pokemon" data-test-id="hidden-pokemon"
            :src="imgSrc" />

        <img alt="Pokémon" class="fade-in" data-test-id="visible-pokemon"
            v-if="showPokemon"
            :src="imgSrc" />
    </div>
</template>

<script>
export default {
    name: 'PokemonPicture',

    props: {
        pokemonId: {
            type: Number,
            required: true,
        },
        showPokemon: {
            type: Boolean,
            required: true,
            default: false,
        }
    },

    computed: {
        imgSrc() {
            return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`;
        }
    }
}
</script>

<style scoped>
.pokemon-container {
    height: 200px;
}
img {
    height: 200px;
    position: absolute;
    width: 100vw;
    margin-left: -50vw;
    user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
}
.hidden-pokemon {
    filter: brightness(0);
}
</style>
