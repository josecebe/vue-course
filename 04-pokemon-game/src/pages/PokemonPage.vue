<template>
    <div v-if="pokemon">
        <h1>Who's that Pok√©mon!?</h1>

        <PokemonPicture
            :pokemon-id="pokemon.id"
            :show-pokemon="showPokemon" />

        <PokemonOptions
            :pokemon-arr="pokemonArr"
            @pokemon-selection="checkAnswer" />

        <template v-if="showAnswer">
            <h2>{{message}}</h2>
            <button @click="newGame">New Game</button>
        </template>
    </div>
    <h1 v-else="pokemon">Wait, please...</h1>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
    name: 'PokemonPage',

    components: {
        PokemonPicture,
        PokemonOptions
    },

    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: null,
        }
    },

    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions();
            const randomInt = Math.floor(Math.random() * 4);
            this.pokemon = this.pokemonArr[randomInt];
        },

        checkAnswer(selectedPokemonId) {
            this.showPokemon = true;
            this.showAnswer = true;

            if (selectedPokemonId === this.pokemon.id) {
                this.message = `Correct! ${this.pokemon.name}`;
            } else {
                this.message = `Oops! ${this.pokemon.name}`;
            }
        },

        newGame() {
            this.pokemonArr = [];
            this.pokemon = null;
            this.showPokemon = false;
            this.showAnswer = false;
            this.message = null;
            this.mixPokemonArray();
        }
    },

    mounted() {
        this.mixPokemonArray();
    }
}
</script>
