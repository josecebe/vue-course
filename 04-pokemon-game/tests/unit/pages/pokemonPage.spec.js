import {shallowMount, mount} from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';
import {pokemonsMock} from '../mocks/pokemons.mock';

describe('PokemonPage tests', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    });

    test('Should match the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Should call mixPokemonArray method on mount', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');

        wrapper = shallowMount(PokemonPage);

        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('Should match the snapshot when pokemons are loaded', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: null,
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Should show PokemonPicture & PokemonOptions components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: null,
                }
            }
        });

        const pokemonPicture = wrapper.findComponent(PokemonPicture);
        const pokemonOptions = wrapper.findComponent(PokemonOptions);

        expect(pokemonPicture.exists()).toBeTruthy();
        expect(pokemonOptions.exists()).toBeTruthy();
        expect(pokemonPicture.props('pokemonId')).toEqual(pokemonsMock[0].id);
        expect(pokemonOptions.props('pokemonsArr')).not.toBeNull();
    });

    test('CheckAnswer method should works correctly', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: null,
                }
            }
        });

        await wrapper.vm.checkAnswer(1);

        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBeTruthy();
        expect(wrapper.vm.message).toBe(`Correct! ${pokemonsMock[0].name}`);

        await wrapper.vm.checkAnswer(2);
        expect(wrapper.vm.message).toBe(`Oops! ${pokemonsMock[0].name}`);
    });
});
