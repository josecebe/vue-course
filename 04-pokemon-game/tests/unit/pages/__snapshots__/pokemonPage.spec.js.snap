// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage tests Should match the snapshot 1`] = `<h1>Wait, please...</h1>`;

exports[`PokemonPage tests Should match the snapshot when pokemons are loaded 1`] = `
<div>
  <h1>Who's that Pok√©mon!?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemonarr="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
